File: ios/PTimer/ExposureCalculator/FilmContext/CustomFilmFormulaGuard.swift
import Foundation

/// Shared correctness helpers for
/// the custom-film formula model. The previous validator only
/// sampled the no-correction boundary, which let `exponent < 1`
/// formulas slip through even though they would shorten metered
/// exposures at longer times. This helper centralises the usable-
/// range check so the editor validate path, the persistence
/// sanitation path, and the preview presenter all enforce the
/// same rule.
enum CustomFilmFormulaGuard {
    /// Upper sample horizon for `Unlimited` valid-through. 3
    /// hours is enough to catch any sub-1 exponent that would
    /// shorten realistic photographic exposures while avoiding
    /// floating-point overflow at extreme metered values.
    static let unlimitedSampleHorizonSeconds: Double = 10800

    /// Sample count for the log-spaced range scan. 24 samples
    /// across [noCorrectionThrough, upper] gives sub-stop
    /// resolution while keeping validation cheap on every
    /// keystroke.
    static let sampleCount: Int = 24

    /// Returns `true` when `Tc(T_m) = baseTc · (T_m / baseTm)^e + offset`
    /// stays at or above `T_m` everywhere in the formula's usable
    /// range. The check is sample-based with a 1 ms tolerance so a
    /// perfectly flat boundary (T_c = T_m at the lower edge) is
    /// not rejected by rounding noise.
    ///
    /// `Unlimited` (`validThrough == nil`) with `exponent < 1` is
    /// rejected outright because the formula will eventually
    /// shorten any metered value past the chosen sample horizon —
    /// trying to draw the line elsewhere produces fragile
    /// edge cases. The same Unlimited case with `exponent == 1`
    /// is caught by the sample at the horizon when
    /// `baseTc < baseTm` (the limit of T_c/T_m is `baseTc / baseTm`).
    /// Parameter bundle for `passesUsableRangeCheck` — kept as a
    /// struct so the helper stays under the swiftlint parameter
    /// limit and call sites can name fields without positional
    /// ambiguity.
    /// Parameter bundle for `passesUsableRangeCheck`. Field names
    /// mirror the shared `ReciprocityFormula` so custom
    /// and preset profiles speak the same vocabulary at the
    /// authoring layer.
    struct UsableRangeInput {
        let exponent: Double
        let referenceMeteredTimeSeconds: Double
        let coefficientSeconds: Double
        let offsetSeconds: Double
        let noCorrectionThroughSeconds: Double
        let sourceRangeThroughSeconds: Double?
    }

    static func passesUsableRangeCheck(_ input: UsableRangeInput) -> Bool {
        let exponent = input.exponent
        let referenceMeteredTime = input.referenceMeteredTimeSeconds
        let coefficient = input.coefficientSeconds
        let offset = input.offsetSeconds
        let noCorrectionThrough = input.noCorrectionThroughSeconds
        let sourceRangeThrough = input.sourceRangeThroughSeconds
        // Defensive: the validator parses for finiteness already,
        // but guard against pathological inputs the sanitation
        // path may receive directly.
        guard exponent.isFinite, referenceMeteredTime.isFinite, referenceMeteredTime > 0,
              coefficient.isFinite, coefficient > 0,
              offset.isFinite,
              noCorrectionThrough.isFinite, noCorrectionThrough >= 0 else {
            return false
        }

        if exponent < 1, sourceRangeThrough == nil {
            return false
        }

        let upper = sourceRangeThrough ?? unlimitedSampleHorizonSeconds
        let lower = max(noCorrectionThrough, 1e-3)
        guard upper > lower else {
            // Range collapses to a single point — only the
            // boundary check matters and `Tc(lower) >= lower` is
            // the right test.
            let tc = coefficient * pow(lower / referenceMeteredTime, exponent) + offset
            return tc + 0.001 >= lower
        }

        let logLower = log(lower)
        let logUpper = log(upper)
        for i in 0..<sampleCount {
            let t = Double(i) / Double(sampleCount - 1)
            let metered = exp(logLower + (logUpper - logLower) * t)
            let tc = coefficient * pow(metered / referenceMeteredTime, exponent) + offset
            if tc + 0.001 < metered {
                return false
            }
        }
        return true
    }
}

/// Parser for the duration-style
/// strings the editor accepts in the application-range fields
/// (`No correction up to`, `Source range through`).
///
/// Accepted shapes:
/// - empty string → `nil` (caller decides default / Unlimited)
/// - plain decimal `"100"` → 100s
/// - suffixed `"100s"`, `"5m"`, `"1h"` → 100s / 300s / 3600s
/// - case-insensitive `"unlimited"` → returns `.unlimited`
/// - anything else → returns `nil`
enum CustomFilmDurationParser {
    enum ParsedDuration: Equatable {
        case seconds(Double)
        case unlimited
        case empty
    }

    static func parse(_ text: String) -> ParsedDuration? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return .empty }
        if trimmed.caseInsensitiveCompare("unlimited") == .orderedSame {
            return .unlimited
        }
        let lowered = trimmed.lowercased()
        if let plain = Double(lowered), plain.isFinite {
            return .seconds(plain)
        }
        if let unit = lowered.last {
            let body = String(lowered.dropLast())
            guard let value = Double(body), value.isFinite else { return nil }
            switch unit {
            case "s": return .seconds(value)
            case "m": return .seconds(value * 60)
            case "h": return .seconds(value * 3600)
            default: return nil
            }
        }
        return nil
    }
}

File: android/core/src/main/kotlin/com/sangwook/ptimer/core/reciprocity/CustomFilmFormulaGuard.kt
package com.sangwook.ptimer.core.reciprocity

import kotlin.math.abs
import kotlin.math.pow

/**
 * Analytic no-shortening guard for custom formula profiles: returns true
 * when no `Tm` in the usable range produces `Tc < Tm`. Mirrors iOS
 * `CustomFilmFormulaGuard` (endpoint + interior-critical-point analysis,
 * not sampling).
 */
object CustomFilmFormulaGuard {
    private const val SLACK_SECONDS: Double = 1e-6

    data class UsableRangeInput(
        val exponent: Double,
        val referenceMeteredTimeSeconds: Double,
        val coefficientSeconds: Double,
        val offsetSeconds: Double,
        val noCorrectionThroughSeconds: Double,
        val sourceRangeThroughSeconds: Double?,
    )

    fun passesUsableRangeCheck(input: UsableRangeInput): Boolean {
        val exponent = input.exponent
        val referenceMeteredTime = input.referenceMeteredTimeSeconds
        val coefficient = input.coefficientSeconds
        val offset = input.offsetSeconds
        val noCorrectionThrough = input.noCorrectionThroughSeconds
        val sourceRangeThrough = input.sourceRangeThroughSeconds

        if (!(exponent.isFinite() && exponent > 0)) return false
        if (!(referenceMeteredTime.isFinite() && referenceMeteredTime > 0)) return false
        if (!(coefficient.isFinite() && coefficient > 0)) return false
        if (!offset.isFinite()) return false
        if (!(noCorrectionThrough.isFinite() && noCorrectionThrough >= 0)) return false
        if (sourceRangeThrough != null &&
            !(sourceRangeThrough.isFinite() && sourceRangeThrough > noCorrectionThrough)
        ) {
            return false
        }

        val scaledCoefficient = coefficient / referenceMeteredTime.pow(exponent)
        val lower = maxOf(noCorrectionThrough, 1e-9)
        val formula = ShorteningFormula(scaledCoefficient, exponent, offset)

        if (abs(exponent - 1.0) < 1e-9) {
            return passesLinearCase(formula, lower, sourceRangeThrough)
        }

        if (exponent < 1) {
            val upper = sourceRangeThrough ?: return false
            return formula.fIsNonNegative(lower) && formula.fIsNonNegative(upper)
        }

        // exponent > 1: convex
        if (!formula.fIsNonNegative(lower)) return false
        if (sourceRangeThrough != null && !formula.fIsNonNegative(sourceRangeThrough)) return false
        val denominator = scaledCoefficient * exponent
        if (denominator <= 0) return true
        val critical = (1.0 / denominator).pow(1.0 / (exponent - 1.0))
        val upperBound = sourceRangeThrough ?: Double.POSITIVE_INFINITY
        if (critical.isFinite() && critical > lower && critical < upperBound) {
            return formula.fIsNonNegative(critical)
        }
        return true
    }

    private fun passesLinearCase(
        formula: ShorteningFormula,
        lower: Double,
        sourceRangeThrough: Double?,
    ): Boolean {
        val slope = formula.coefficient - 1.0
        if (slope > 1e-9) return formula.fIsNonNegative(lower)
        if (slope < -1e-9) {
            val upper = sourceRangeThrough ?: return false
            return formula.fIsNonNegative(lower) && formula.fIsNonNegative(upper)
        }
        return formula.offset >= -SLACK_SECONDS
    }

    private class ShorteningFormula(
        val coefficient: Double,
        val exponent: Double,
        val offset: Double,
    ) {
        fun fIsNonNegative(t: Double): Boolean {
            if (t <= 0) return offset >= -SLACK_SECONDS
            val tc = coefficient * t.pow(exponent) + offset
            return tc >= t - SLACK_SECONDS
        }
    }
}
